lib/internal/modules/esm/resolve.js

```diff
@@ -59,6 +60,36 @@ const userConditions = getOptionValue('--conditions');
 const DEFAULT_CONDITIONS = ObjectFreeze(['node', 'import', ...userConditions]);
 const DEFAULT_CONDITIONS_SET = new SafeSet(DEFAULT_CONDITIONS);
 
+const pendingDeprecation = getOptionValue('--pending-deprecation');
+const emittedPackageWarnings = new SafeSet();
+function emitFolderMapDeprecation(match, pjsonUrl, isExports, base) {
+  const pjsonPath = fileURLToPath(pjsonUrl);
+  if (!pendingDeprecation) {
+    const nodeModulesIndex = StringPrototypeLastIndexOf(pjsonPath,
+                                                        '/node_modules/');
+    if (nodeModulesIndex !== -1) {
+      const afterNodeModulesPath = StringPrototypeSlice(pjsonPath,
+                                                        nodeModulesIndex + 14,
+                                                        -13);
+      try {
+        const { packageSubpath } = parsePackageName(afterNodeModulesPath);
+        if (packageSubpath === '.')
+          return;
+      } catch {}
+    }
+  }
+  if (emittedPackageWarnings.has(pjsonPath + '|' + match))
+    return;
+  emittedPackageWarnings.add(pjsonPath + '|' + match);
+  process.emitWarning(
+    `Use of deprecated folder mapping "${match}" in the ${isExports ?
+      '"exports"' : '"imports"'} field module resolution of the package at ${
+      pjsonPath}${base ? ` imported from ${fileURLToPath(base)}` : ''}.\n` +
+      `Update this package.json to use a subpath pattern like "${match}*".`,
```

Comment:
would it be reasonable in this message to mention which node versions support subpath patterns, so package authors don't unknowingly make a breaking change?

Reply:
We have version information in the documentation for this under the exports field.

Keeping it up to date here might be tricky though, especially with a backport still in progress.

It's a good point though - we could possibly delay this landing until the 12 backport is released if you'd prefer.

Reply:
I don't think we should print version information here. It should be available in the docs and to the best of my knowledge we d not have that information available in any other deprecation warnings

Reply:
Fair. How likely are other deprecation warnings to cause package breakage compared to this one? Maybe we could at least say "Please check the docs for which node versions support subpath patterns" and link them there?

Reply:
Subpath patterns are on 14.13, 15 and will be backported to 12 very soon so I'm just not sure what value there is in muddying the advice for users here? The deprecation has a clear action and that is to move away from the feature or upgrade to subpaths.

Reply:
Just because node doesn’t support a version doesn’t mean package authors don’t. It’s obv fine if you don’t want to take my suggestion, but i think the risk of causing breakage is far greater than the risk of someone remaining on slash patterns due to an extra sentence.

Reply:
Agreed there is a risk, but my concern is that any advice with respect to being cautious about subpaths support will be outdated by the time most users read it.

Even if we weren't backporting to Node.js 12, a deprecation of a feature on 15/16 that points to using a feature only available on 14 seems fine to me and I'm not sure there is precedent for warning users from using a feature (from the web to Node.js) just because it's not available 3 versions back.

Reply:
(within the platform warning itself)